File: src/Python/devpossible_ton/devpossible_ton/serializer/ton_serializer.py
"""
TonSerializer - Serializer for TON format
Copyright (c) 2024 DevPossible, LLC
"""

from typing import Optional, Any
from ..models import TonDocument, TonObject, TonValue, TonArray, TonEnum
from .ton_serialize_options import TonSerializeOptions, TonFormatStyle


class TonSerializer:
    """Serializer for TON format."""

    def __init__(self, options: Optional[TonSerializeOptions] = None):
        self.options = options or TonSerializeOptions()

    def serialize(self, document: TonDocument) -> str:
        """Serialize a document to TON format."""
        result = []
        
        # Add header if requested
        if self.options.include_header:
            result.append(f"#TON {self.options.ton_version}")
            if self.options.schema_file:
                result.append(f"#SCHEMA {self.options.schema_file}")
            result.append("")
        
        # Serialize the root value
        result.append(self._serialize_value(document.get_root(), 0))
        
        return self.options.line_ending.join(result)

    def _serialize_value(self, value: Any, indent_level: int) -> str:
        """Serialize a value."""
        if isinstance(value, TonObject):
            return self._serialize_object(value, indent_level)
        elif isinstance(value, dict):
            # Handle plain Python dicts by wrapping in TonObject
            obj = TonObject()
            for k, v in value.items():
                obj.set(k, v)
            return self._serialize_object(obj, indent_level)
        elif isinstance(value, TonArray):
            return self._serialize_array(value, indent_level)
        elif isinstance(value, list):
            # Handle plain Python lists by wrapping in TonArray
            arr = TonArray()
            for item in value:
                arr.add(item)
            return self._serialize_array(arr, indent_level)
        elif isinstance(value, TonValue):
            # Handle enum and enumSet type hints
            if hasattr(value, 'type_hint') and value.type_hint:
                if value.type_hint == 'enum':
                    return f'|{value.get_value()}|'
                elif value.type_hint == 'enumSet':
                    val = value.get_value()
                    if isinstance(val, list):
                        return '|' + '|'.join(str(v) for v in val) + '|'
                    return f'|{val}|'
                else:
                    return self._serialize_primitive(value.get_value(), value.type_hint)
            return self._serialize_primitive(value.get_value())
        elif isinstance(value, TonEnum):
            return str(value)
        else:
            return self._serialize_primitive(value)

    def _serialize_object(self, obj: TonObject, indent_level: int) -> str:
        """Serialize an object."""
        is_compact = self.options.format_style == TonFormatStyle.COMPACT
        
        # Add class name prefix if present
        prefix = ''
        if obj.class_name:
            prefix = obj.class_name
            if obj.instance_count is not None:
                prefix += f'({obj.instance_count})'
        
        # Check if object is empty (or all values are omitted)
        non_omitted_items = []
        keys = sorted(obj.keys()) if self.options.sort_properties else obj.keys()
        for key in keys:
            # Skip metadata properties (_className, _instanceCount)
            if key.startswith('_'):
                continue
            value = obj.get(key)
            if not self._should_omit_value(value):
                non_omitted_items.append((key, value))
        
        # If empty, return compact format regardless of style
        if len(non_omitted_items) == 0:
            return prefix + '{}'
        
        if is_compact:
            items = []
            for key, value in non_omitted_items:
                val_str = self._serialize_value(value, 0)
                items.append(f'{key}{self.options.property_separator}{val_str}')
            content = '{' + ','.join(items) + '}'
            return prefix + content
        else:
            lines = ['{']
            indent = self.options.indentation or ''
            current_indent = indent * (indent_level + 1)
            
            for key, value in non_omitted_items:
                val_str = self._serialize_value(value, indent_level + 1)
                lines.append(f'{current_indent}{key}{self.options.property_separator}{val_str}')
            
            lines.append((indent * indent_level) + '}')
            content = self.options.line_ending.join(lines)
            return prefix + content

    def _serialize_array(self, arr: TonArray, indent_level: int) -> str:
        """Serialize an array."""
        if self.options.omit_empty_collections and arr.length() == 0:
            return '[]'
        
        is_compact = self.options.format_style == TonFormatStyle.COMPACT
        items = [self._serialize_value(item, indent_level) for item in arr.to_array()]
        
        if is_compact:
            return '[' + self.options.array_separator.join(items) + ']'
        else:
            indent = self.options.indentation or ''
            current_indent = indent * (indent_level + 1)
            # Pretty arrays don't have commas between lines (based on test expectations)
            return '[' + self.options.line_ending + \
                   self.options.line_ending.join(f'{current_indent}{item}' for item in items) + \
                   self.options.line_ending + (indent * indent_level) + ']'

    def _serialize_primitive(self, value: Any, type_hint: Optional[str] = None) -> str:
        """Serialize a primitive value."""
        if value is None:
            return 'null'
        elif value is True:
            # Handle type hints for booleans
            if self.options.include_type_hints and type_hint:
                prefix = self._get_type_hint_prefix(type_hint)
                return f'{prefix}true'
            return 'true'
        elif value is False:
            # Handle type hints for booleans
            if self.options.include_type_hints and type_hint:
                prefix = self._get_type_hint_prefix(type_hint)
                return f'{prefix}false'
            return 'false'
        elif isinstance(value, str):
            # Check if it's a GUID
            if self._is_guid(value):
                return value
            
            # Handle type hints with prefixes
            if self.options.include_type_hints and type_hint:
                prefix = self._get_type_hint_prefix(type_hint)
                quote = '"' if self.options.quote_style == 'double' else "'"
                return f'{prefix}{quote}{value}{quote}'
            
            quote = '"' if self.options.quote_style == 'double' else "'"
            # Escape quotes in the string
            escaped = value.replace('\\', '\\\\').replace(quote, f'\\{quote}')
            return f'{quote}{escaped}{quote}'
        elif hasattr(value, 'isoformat'):
            # Handle datetime objects
            iso_string = value.isoformat()
            if 'T' in iso_string:
                # Convert to date-only format if it's midnight
                iso_string = iso_string.split('T')[0] if value.hour == 0 and value.minute == 0 and value.second == 0 else iso_string
            
            if self.options.include_type_hints and type_hint:
                prefix = self._get_type_hint_prefix(type_hint)
                quote = '"' if self.options.quote_style == 'double' else "'"
                return f'{prefix}{quote}{iso_string}{quote}'
            
            quote = '"' if self.options.quote_style == 'double' else "'"
            return f'{quote}{iso_string}{quote}'
        else:
            # Handle type hints for numbers
            if self.options.include_type_hints and type_hint:
                prefix = self._get_type_hint_prefix(type_hint)
                return f'{prefix}{value}'
            return str(value)
    
    def _is_guid(self, value: str) -> bool:
        """Check if a string is a GUID."""
        import re
        guid_pattern = r'^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$'
        return bool(re.match(guid_pattern, value))
    
    def _get_type_hint_prefix(self, type_hint: str) -> str:
        """Get the prefix for a type hint."""
        prefixes = {
            'string': '$',
            'number': '%',
            'boolean': '&',
            'date': '^'
        }
        return prefixes.get(type_hint.lower(), '')
    
    def _should_omit_value(self, value: Any) -> bool:
        """Check if a value should be omitted based on options."""
        # Check for TonValue wrapping
        actual_value = value.get_value() if hasattr(value, 'get_value') else value
        
        if actual_value is None:
            # Only omit if explicitly requested
            if self.options.omit_nulls:
                return True
            # Don't auto-omit nulls even if omit_undefined is true
            # (in TON, null and undefined are different)
            return False
        
        if isinstance(value, (TonArray, list)) and len(value) == 0 and self.options.omit_empty_collections:
            return True
        if isinstance(value, (TonObject, dict)) and len(value) == 0 and self.options.omit_empty_collections:
            return True
        return False

File: src/JavaScript/devpossible-ton/src/serializer/TonSerializer.ts
/**
 * TonSerializer - Serializer for TON format
 * Copyright (c) 2024 DevPossible, LLC
 */

import { TonDocument } from '../models/TonDocument';
import { TonObject } from '../models/TonObject';
import { TonValue } from '../models/TonValue';
import { TonArray } from '../models/TonArray';
import { TonEnum } from '../models/TonEnum';
import { TonSerializeOptions, ITonSerializeOptions, TonFormatStyle } from './TonSerializeOptions';

// Re-export for convenience
export { TonSerializeOptions, TonFormatStyle };

export class TonSerializer {
  private defaultOptions: TonSerializeOptions;

  constructor(options?: TonSerializeOptions | ITonSerializeOptions) {
    // If options is a TonSerializeOptions instance, use it directly
    // Otherwise, create a new instance with the provided options
    if (options instanceof TonSerializeOptions) {
      this.defaultOptions = options;
    } else {
      this.defaultOptions = new TonSerializeOptions(options);
    }
  }

  /**
   * Get the effective options for a serialize call
   */
  private getOptions(options?: TonSerializeOptions | ITonSerializeOptions): TonSerializeOptions {
    if (options instanceof TonSerializeOptions) {
      return options;
    } else if (options) {
      return new TonSerializeOptions(options);
    }
    return this.defaultOptions;
  }

  /**
   * Serializes a TonDocument or object to TON format string
   * Matches C# signature: Serialize(object obj, TonSerializeOptions? options = null)
   */
  public serialize(obj: any, options?: TonSerializeOptions | ITonSerializeOptions): string {
    const opts = this.getOptions(options);

    // Temporarily set options for this serialize call
    const savedOptions = this.defaultOptions;
    this.defaultOptions = opts;

    try {
      if (obj instanceof TonDocument) {
        return this._serializeDocument(obj);
      } else if (obj instanceof TonObject) {
        return this.serializeObject(obj, 0);
      } else if (obj instanceof TonArray) {
        return this.serializeArray(obj, 0);
      } else {
        // Convert plain object to TonDocument
        const doc = TonDocument.fromObject(obj);
        return this._serializeDocument(doc);
      }
    } finally {
      // Restore default options
      this.defaultOptions = savedOptions;
    }
  }

  /**
   * Serializes a TonDocument to string
   */
  public serializeDocument(doc: TonDocument, options?: TonSerializeOptions | ITonSerializeOptions): string {
    const savedOptions = this.defaultOptions;
    try {
      const opts = this.getOptions(options);
      this.defaultOptions = opts;
      return this._serializeDocument(doc);
    } finally {
      this.defaultOptions = savedOptions;
    }
  }

  private _serializeDocument(doc: TonDocument): string {
    const parts: string[] = [];

    // Add header if requested
    if (this.defaultOptions.includeHeader && this.defaultOptions.tonVersion) {
      parts.push(`#@ tonVersion = '${this.defaultOptions.tonVersion}'`);
      parts.push('');
    }

    // Serialize root object
    if (doc.root instanceof TonObject) {
      parts.push(this.serializeObject(doc.root, 0));
    } else if (doc.root instanceof TonArray) {
      parts.push(this.serializeArray(doc.root, 0));
    } else if (doc.root instanceof TonValue) {
      parts.push(this.serializeValue(doc.root, 0));
    } else {
      parts.push(this.serializePlainValue(doc.root, 0));
    }

    return parts.join(this.defaultOptions.lineEnding || '\n');
  }

  /**
   * Serializes a TonObject to string
   */
  private serializeObject(obj: TonObject, indent: number): string {
    const props = obj.properties;

    // Handle empty object
    if (props.size === 0) {
      // Check if it has a className
      if (obj.className) {
        const prefix = obj.instanceCount !== undefined ?
          `(${obj.className})(${obj.instanceCount})` :
          `(${obj.className})`;
        return `${prefix}{}`;
      }
      return '{}';
    }

    // Sort properties if requested
    let keys = Array.from(props.keys());
    if (this.defaultOptions.sortProperties) {
      keys.sort();
    }

    // Filter out nulls/undefined if requested
    keys = keys.filter(key => {
      const value = props.get(key);
      const actualValue = value instanceof TonValue ? value.value : value;
      if (this.defaultOptions.omitNulls && actualValue === null) return false;
      if (this.defaultOptions.omitUndefined && actualValue === undefined) return false;
      return true;
    });

    // Format based on style
    if (this.defaultOptions.formatStyle === TonFormatStyle.Compact) {
      return this.serializeObjectCompact(obj, keys);
    } else {
      return this.serializeObjectPretty(obj, keys, indent);
    }
  }

  /**
   * Serializes an object in compact format
   */
  private serializeObjectCompact(obj: TonObject, keys: string[]): string {
    // Add class name if present
    let prefix = '';
    if (obj.className) {
      prefix = obj.instanceCount !== undefined ?
        `(${obj.className})(${obj.instanceCount})` :
        `(${obj.className})`;
    }

    const separator = this.defaultOptions.propertySeparator || ' = ';
    const propParts = keys.map(key => {
      const value = obj.properties.get(key)!;
      const keyStr = this.serializePropertyName(key);
      const valueStr = this.serializeAnyValue(value, 0);
      return `${keyStr}${separator}${valueStr}`;
    });

    const joinStr = separator === ':' ? ',' : ', ';
    return `${prefix}{${propParts.join(joinStr)}}`;
  }

  /**
   * Serializes an object in pretty format
   */
  private serializeObjectPretty(obj: TonObject, keys: string[], indent: number): string {
    const indentStr = this.getIndentString(indent);
    const nextIndentStr = this.getIndentString(indent + 1);
    const parts: string[] = [];

    // Add class name if present
    let prefix = '';
    if (obj.className) {
      prefix = obj.instanceCount !== undefined ?
        `(${obj.className})(${obj.instanceCount})` :
        `(${obj.className})`;
    }

    parts.push(`${prefix}{`);

    const separator = this.defaultOptions.propertySeparator || ' = ';
    keys.forEach((key, index) => {
      const value = obj.properties.get(key)!;
      const keyStr = this.serializePropertyName(key);
      const valueStr = this.serializeAnyValue(value, indent + 1);
      // Add comma after each property except the last (unless trailingCommas is true)
      const isLast = index === keys.length - 1;
      const comma = (this.defaultOptions.trailingCommas || !isLast) ? ',' : '';
      parts.push(`${nextIndentStr}${keyStr}${separator}${valueStr}${comma}`);
    });

    parts.push(`${indentStr}}`);

    return parts.join(this.defaultOptions.lineEnding);
  }

  /**
   * Serializes a property name
   */
  private serializePropertyName(name: string): string {
    // Check if it has a type annotation (name:type format)
    const typeAnnotationMatch = name.match(/^([a-zA-Z_][a-zA-Z0-9_]*):([a-zA-Z_][a-zA-Z0-9_]*)$/);
    if (typeAnnotationMatch) {
      return name; // Return as-is for type annotations
    }

    // Check if it's a valid identifier
    if (/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(name) || /^[0-9]+(\.[0-9]+)?$/.test(name)) {
      return name;
    }
    // Otherwise quote it
    return this.quoteString(name);
  }

  /**
   * Serializes a TonArray to string
   */
  private serializeArray(arr: TonArray, indent: number): string {
    const items = arr.items;

    if (items.length === 0) {
      return '[]';
    }

    // In Pretty mode, format root-level arrays with indentation
    // Nested arrays stay inline
    if (this.defaultOptions.formatStyle === TonFormatStyle.Pretty && indent === 0) {
      const indentStr = this.getIndentString(indent);
      const nextIndentStr = this.getIndentString(indent + 1);
      const parts: string[] = [];

      parts.push('[');

      items.forEach((item, index) => {
        const valueStr = this.serializeAnyValue(item, indent + 1);
        // Add comma after each item except the last (unless trailingCommas is true)
        const isLast = index === items.length - 1;
        const comma = (this.defaultOptions.trailingCommas || !isLast) ? ',' : '';
        parts.push(`${nextIndentStr}${valueStr}${comma}`);
      });

      parts.push(`${indentStr}]`);

      return parts.join(this.defaultOptions.lineEnding);
    } else {
      // Nested arrays and compact mode: stay inline
      const itemStrs = items.map(item => this.serializeAnyValue(item, indent));
      // Use arraySeparator if specified, otherwise default based on format style
      const separator = this.defaultOptions.arraySeparator !== undefined
        ? this.defaultOptions.arraySeparator
        : (this.defaultOptions.formatStyle === TonFormatStyle.Compact ? ',' : ', ');
      return `[${itemStrs.join(separator)}]`;
    }
  }

  /**
   * Serializes any value
   */
  private serializeAnyValue(value: any, indent: number): string {
    if (value instanceof TonObject) {
      return this.serializeObject(value, indent);
    } else if (value instanceof TonArray) {
      return this.serializeArray(value, indent);
    } else if (value instanceof TonValue) {
      return this.serializeValue(value, indent);
    } else if (value instanceof TonEnum) {
      return this.serializeEnum(value);
    } else {
      return this.serializePlainValue(value, indent);
    }
  }

  /**
   * Serializes a TonValue
   */
  private serializeValue(value: TonValue, indent: number): string {
    // Handle enum and enumSet specially
    if (value.typeHint === 'enum') {
      if (Array.isArray(value.value)) {
        return '|' + value.value.join('|') + '|';
      } else {
        return '|' + value.value + '|';
      }
    }
    if (value.typeHint === 'enumSet') {
      if (Array.isArray(value.value)) {
        return '|' + value.value.join('|') + '|';
      } else {
        return '|' + value.value + '|';
      }
    }

    // If the wrapped value is a TonObject or TonArray, serialize it properly
    if (value.value instanceof TonObject) {
      return this.serializeObject(value.value, indent);
    }
    if (value.value instanceof TonArray) {
      return this.serializeArray(value.value, indent);
    }

    let result = this.serializePlainValue(value.value, indent);

    // Add type hint if requested and present
    if (this.defaultOptions.includeTypeHints && value.typeHint) {
      const hint = this.getTypeHintPrefix(value.typeHint);
      if (hint) {
        result = hint + result;
      }
    }

    return result;
  }

  /**
   * Serializes a TonEnum
   */
  private serializeEnum(enumValue: TonEnum): string {
    if (Array.isArray(enumValue.values)) {
      // EnumSet
      return '|' + enumValue.values.join('|') + '|';
    } else {
      // Single enum
      return '|' + enumValue.values + '|';
    }
  }

  /**
   * Serializes a plain JavaScript value
   */
  private serializePlainValue(value: any, indent: number): string {
    if (value === null) {
      return 'null';
    }
    if (value === undefined) {
      return 'undefined';
    }
    if (typeof value === 'boolean') {
      return value ? 'true' : 'false';
    }
    if (typeof value === 'number') {
      // Handle special numbers
      if (Number.isNaN(value)) return 'NaN';
      if (value === Infinity) return 'Infinity';
      if (value === -Infinity) return '-Infinity';

      // Check if it's an integer and within safe range
      if (Number.isInteger(value) && Math.abs(value) <= Number.MAX_SAFE_INTEGER) {
        return value.toString();
      }
      // Use exponential notation for very large/small numbers
      if (Math.abs(value) >= 1e21 || (Math.abs(value) < 1e-6 && value !== 0)) {
        return value.toExponential();
      }
      return value.toString();
    }
    if (typeof value === 'string') {
      // Check if it's a GUID (with or without braces)
      const guidPattern = /^(\{)?[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(\})?$/i;
      if (guidPattern.test(value)) {
        return value;
      }
      return this.quoteString(value);
    }
    if (value instanceof Date) {
      return this.quoteString(value.toISOString());
    }
    if (Array.isArray(value)) {
      // Convert to TonArray and serialize
      const tonArray = new TonArray();
      tonArray.items = value;
      return this.serializeArray(tonArray, indent);
    }
    if (typeof value === 'object') {
      // Convert to TonObject and serialize
      const tonObj = new TonObject();
      for (const [key, val] of Object.entries(value)) {
        tonObj.set(key, val);
      }
      return this.serializeObject(tonObj, indent);
    }

    return String(value);
  }

  /**
   * Quotes a string value
   */
  private quoteString(str: string): string {
    const quote = this.defaultOptions.quoteStyle === 'single' ? "'" : '"';
    const escaped = str
      .replace(/\\/g, '\\\\')
      .replace(new RegExp(quote, 'g'), '\\' + quote)
      .replace(/\n/g, '\\n')
      .replace(/\r/g, '\\r')
      .replace(/\t/g, '\\t');

    // Check if it's a multiline string
    if (str.includes('\n') && this.defaultOptions.formatStyle === TonFormatStyle.Pretty) {
      // Use triple quotes for multiline
      return `"""${str}"""`;
    }

    return quote + escaped + quote;
  }

  /**
   * Gets the type hint prefix character
   */
  private getTypeHintPrefix(typeHint: string): string {
    switch (typeHint) {
      case 'string': return '$';
      case 'number': return '%';
      case 'boolean': return '&';
      case 'date': return '^';
      default: return '';
    }
  }

  /**
   * Gets the indentation string for a given level
   */
  private getIndentString(level: number): string {
    if (this.defaultOptions.formatStyle === TonFormatStyle.Compact) {
      return '';
    }
    return (this.defaultOptions.indentChar || ' ').repeat((this.defaultOptions.indentSize || 4) * level);
  }
}

// Static methods for convenience
export namespace TonSerializer {
  export function stringify(obj: any, options?: TonSerializeOptions): string {
    const serializer = new TonSerializer(options);
    return serializer.serialize(obj);
  }

  export function compact(obj: any): string {
    return stringify(obj, { formatStyle: TonFormatStyle.Compact });
  }

  export function pretty(obj: any): string {
    return stringify(obj, { formatStyle: TonFormatStyle.Pretty });
  }
}